File: python/interlock_fastapi/sink.py
import os
import time
from typing import Optional

class IncidentSink:
    def __init__(self, log_file: str):
        self.log_file = log_file
        self.ensure_header()

    def ensure_header(self):
        if not os.path.exists(self.log_file):
            os.makedirs(os.path.dirname(self.log_file), exist_ok=True)
            with open(self.log_file, 'w', encoding='utf-8') as f:
                f.write("# Live Incident Log: Interlock Middleware (Python)\n\n")
                f.write("> **Status**: ACTIVE MONITORING\n")
                f.write(f"> **Location**: {self.log_file}\n\n")
                f.write("---\n")
                f.write("## Incident History\n\n")

    def log_event(self, event: dict):
        """
        Logs an event compliant with the Standard Incident Log format.
        """
        timestamp = time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())
        incident_id = event.get("incident_id", "000")
        
        # Determine if Header (New Incident) or Event (Resolution/Update)
        # Using simple heuristic: Is it a Refusal (Start) or Recovery (End)?
        # For remote client, we might just append events linearly.
        
        with open(self.log_file, 'a', encoding='utf-8') as f:
            if "refusal" in str(event.get("trigger", "")).lower():
                 f.write(f"\n### Incident #{incident_id}: Traffic Refusal\n")
                 f.write(f"- **Failure Class**: {event.get('failure_class', 'Unknown')}\n")
                 f.write(f"- **Incident Window**: {timestamp} -> ...\n")
                 f.write(f"- **Trigger**: {event.get('trigger', 'Unknown')}\n\n")
            else:
                 f.write(f"#### Event {incident_id}-X\n")
                 f.write(f"- **Timestamp**: {timestamp}\n")
                 f.write(f"- **Action**: {event.get('action', 'Unknown')}\n")
                 f.write(f"- **Outcome**: {event.get('details', '')}\n")
                 f.write(f"- **Confidence**: {event.get('confidence', 0.0)}\n\n")

File: packages/interlock-express/src/sink.ts
import * as fs from 'fs';
import * as path from 'path';

/**
 * Standard Incident Sink Interface
 * Ensures compliance with docs/LIVE_INCIDENTS.md format
 */

export interface InterlockEvent {
    incidentId?: string; // If part of ongoing incident
    trigger: string;
    action: string;
    details: string;
    recoveryTime: number; // Seconds
    confidence: number;
    failureClass: string;
}

export interface IncidentSink {
    logEvent(event: InterlockEvent): void;
}

// Default Legend to enforce standard compliance
const DEFAULT_LEGEND = `
## Confidence Interpretation
- **≥0.8**: High certainty (Normal operation)
- **0.5–0.79**: Moderate certainty (Protective mode preferred)
- **<0.5**: Low certainty (Refusal required)
`;

export class FileIncidentSink implements IncidentSink {
    private logPath: string;
    private incidentCounter: number = 0;
    private lastLoggedDate: string = '';

    constructor(absolutePath: string) {
        this.logPath = absolutePath;
        this.ensureHeader();
    }

    private ensureHeader() {
        if (!fs.existsSync(this.logPath)) {
            const header = `# Live Incident Log: Interlock Middleware

> **Status**: ACTIVE MONITORING
> **Service**: Express Middleware Integration
> **Location**: ${this.logPath}

---
${DEFAULT_LEGEND}

## Incident History

`;
            fs.mkdirSync(path.dirname(this.logPath), { recursive: true });
            fs.writeFileSync(this.logPath, header);
        }
    }

    logEvent(event: InterlockEvent): void {
        const timestamp = new Date().toISOString();
        let entry = '';

        if (event.incidentId && !event.incidentId.includes('-')) {
            // It's a header / new incident start
            entry += `
### Incident #${event.incidentId}: ${event.action}
- **Failure Class**: ${event.failureClass}
- **Incident Window**: ${timestamp} → ...
- **Trigger**: ${event.trigger}

`;
        } else {
            // It's an event
            const evtId = event.incidentId || '000-X';
            entry += `#### Event ${evtId}
- **Timestamp**: ${timestamp}
- **Action**: ${event.action}
- **Outcome**: ${event.details}
- **Recovery time**: ${event.recoveryTime.toFixed(1)}s
- **Confidence**: ${event.confidence.toFixed(2)} (Moderate)

`;
        }

        fs.appendFileSync(this.logPath, entry);
    }
}
